File: backend/app/schemas/stock.py
from datetime import datetime
from typing import Literal
from pydantic import BaseModel, Field, model_validator


StockRequestStatus = Literal["pending", "completed", "archived"]


class StockRequestItemCreate(BaseModel):
    product_name: str = Field(min_length=1, max_length=255)
    quantity: int = Field(ge=1)


class StockRequestCreate(BaseModel):
    condominio_id: str | None = None
    product_name: str | None = Field(default=None, min_length=1, max_length=255)
    quantity: int | None = Field(default=None, ge=1)
    photo_name: str | None = Field(default=None, max_length=255)
    photo_data: str | None = None
    items: list[StockRequestItemCreate] | None = Field(default=None, min_length=1)

    @model_validator(mode="after")
    def validate_single_or_batch(self) -> "StockRequestCreate":
        if self.items:
            return self

        if self.product_name is None or self.quantity is None:
            raise ValueError("Provide a product or at least one stock request item")

        return self


class StockRequestStatusUpdate(BaseModel):
    status: StockRequestStatus


class StockRequestRead(BaseModel):
    id: str
    product_name: str
    quantity: int
    photo_name: str | None = None
    photo_data: str | None = None
    status: StockRequestStatus
    created_at: datetime
    updated_at: datetime
    condominio_id: str


class StockRequestCreateResponse(BaseModel):
    data: list[StockRequestRead]
    count: int

File: frontend/src/services/stock.ts
import { api } from "./api";

export type StockRequestStatus = "pending" | "completed" | "archived";

export type StockRequest = {
  id: string;
  product_name: string;
  quantity: number;
  photo_name: string | null;
  photo_data: string | null;
  status: StockRequestStatus;
  created_at: string;
  updated_at: string;
  condominio_id: string;
};

export type StockRequestList = { data: StockRequest[]; count: number };

export const stockService = {
  async create(payload: { product_name: string; quantity: number; photo_name?: string | null; photo_data?: string | null }) {
    const { data } = await api.post<StockRequest>("/stock-requests", payload);
    return data;
  },
  async list(params: { search?: string; status?: StockRequestStatus | ""; date_from?: string; date_to?: string; limit?: number } = {}) {
    const { data } = await api.get<StockRequestList>("/stock-requests", {
      params: { skip: 0, limit: params.limit ?? 200, ...params, status: params.status || undefined }
    });
    return data;
  },
  async complete(id: string) {
    const { data } = await api.patch<StockRequest>(`/stock-requests/${id}/status`, { status: "completed" });
    return data;
  },
  async archive(id: string) {
    const { data } = await api.delete<StockRequest>(`/stock-requests/${id}`);
    return data;
  }
};
